crd-generator/api/src/main/java/io/fabric8/crd/generator/InternalSchemaSwaps.java

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.crd.generator;
+
+import io.sundr.model.ClassRef;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.StringJoiner;
+import java.util.stream.Stream;
+
+public class InternalSchemaSwaps {
+  private final Map<Key, Value> swaps = new HashMap<>();
+
+  public void registerSwap(ClassRef definitionType, ClassRef originalType, String fieldName, ClassRef targetType) {
+    Value value = new Value(definitionType, originalType, fieldName, targetType);
+    swaps.put(new Key(originalType, fieldName), value);
+  }
+
+  public Stream<Value> getUnusedSwaps() {
+    return swaps.values().stream().filter(value -> !value.used);
+  }
+
+  public Optional<ClassRef> lookupAndMark(ClassRef originalType, String name) {
+    Value value = swaps.get(new Key(originalType, name));
+    if (value != null) {
+      value.markUsed();
+      return Optional.of(value.getTargetType());
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  private static class Key {
+    private final ClassRef originalType;
+    private final String fieldName;
+
+    public Key(ClassRef originalType, String fieldName) {
+      this.originalType = originalType;
+      this.fieldName = fieldName;
+    }
+
+    public ClassRef getOriginalType() {
+      return originalType;
+    }
+
+    public String getFieldName() {
+      return fieldName;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      Key key = (Key) o;
+      return Objects.equals(originalType, key.originalType) && Objects.equals(fieldName, key.fieldName);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(originalType, fieldName);
+    }
+
+    @Override
+    public String toString() {
+      return new StringJoiner(", ", Key.class.getSimpleName() + "[", "]")
+          .add("originalType=" + originalType)
+          .add("fieldName='" + fieldName + "'")
+          .toString();
+    }
+  }
+
+  public static class Value {
```

Comment:
This `Value` class spills out, either we give it a more meaningful name, or we contain it making it `private` and returning a more generic type from `getUnusedSwaps` (e.g. directly the resulting `String` to be used in the error message?)

Also,  given this implementation, `Value` can extend `Key`, I believe.

Reply:
I renamed it to `InternalSchemaSwap`.
About extending `Key`, yeah, it's a possibility, but I would rather not do it, because `Key` is an immutable value class, designed to be used as, well, a HashMap key. That has implications on it's `equals` and `hashCode`.
`Value` (or now `InternalSchemaSwap`) is mutable, so either its `equals/hashCode` would have to only use a subset of fields or it could never be used as a map key - but being a subclass of `Key`, any `HashMap<Key, ...>` would allow that. So I chose to not deal with this at all and make it separate.

Reply:
fair, but mark `Key` as `final` please.